File: leetcode/122.买卖股票的最佳时机II/main.py
# coding=utf-8

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0

        for i in range(len(prices) - 1):
            tmp_p = prices[i + 1] - prices[i]

            if tmp_p > 0:
                profit += tmp_p

        return profit


if __name__ == "__main__":
    prices = [7, 1, 5, 3, 6, 4]
    print(Solution().maxProfit(prices))

File: leetcode/122.买卖股票的最佳时机II/main.go
package main

/*
Refer: https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-ii/solutions/476997/shou-hua-tu-jie-xiang-jie-duo-chong-jie-fa-mai-mai/?envType=study-plan-v2&envId=top-interview-150
*/

func maxProfit(prices []int) int {
	lenP := len(prices)
	i := 1
	profit := 0

	for i < lenP {
		// 递减
		for i < lenP && prices[i-1] >= prices[i] {
			i++
		}
		// 到达局部波谷
		trough := prices[i-1]

		// 递增
		for i < lenP && prices[i-1] <= prices[i] {
			i++
		}
		// 到达局部波峰
		peak := prices[i-1]
		profit += peak - trough
	}

	return profit
}

func main() {
	prices := []int{3, 1, 2, 0, 4, 3}
	println(maxProfit(prices))
}
